### ISSUE TITLE
Libewf dependency problem

### ISSUE THREAD

**Issue by kamwoods** [simsong/bulk_extractor#78]
The build instructions for the libewf dependency (currently at https://github.com/simsong/bulk_extractor/wiki/Installing-bulk_extractor) do not appear to be appropriate when working with current libewf-experimental installs.

In an environment where libewf-experimental (checkout from the master) builds and installs cleanly (for reference, this is Ubuntu 16.04 with gcc 5.3.1, but the same behavior can be seen in other environments, including stock Ubuntu 14.04), the bulk_extractor build fails with the following errors:

---

image_process.cpp: In destructor ‘virtual process_ewf::~process_ewf()’:
image_process.cpp:333:21: error: ‘libewf_close’ was not declared in this scope
  libewf_close(handle);
                     ^
image_process.cpp: In member function ‘virtual int process_ewf::open()’:
image_process.cpp:456:98: error: too few arguments to function ‘int libewf_glob(const char_, size_t, uint8_t, char_**, int*, libewf_error_t**)’
     amount_of_filenames = libewf_glob(fname,strlen(fname),LIBEWF_FORMAT_UNKNOWN,&libewf_filenames);
                                                                                                  ^
In file included from image_process.h:207:0,
                 from image_process.cpp:21:
/usr/local/include/libewf.h:134:5: note: declared here
 int libewf_glob(
     ^
image_process.cpp:460:79: error: ‘libewf_open’ was not declared in this scope
     handle = libewf_open(libewf_filenames,amount_of_filenames,LIBEWF_OPEN_READ);
                                                                               ^
image_process.cpp:468:59: error: ‘libewf_get_media_size’ was not declared in this scope
     libewf_get_media_size(handle,(size64_t _)&ewf_filesize);
                                                           ^
image_process.cpp: In member function ‘virtual int process_ewf::pread(unsigned char_, size_t, int64_t) const’:
image_process.cpp:520:54: error: ‘libewf_read_random’ was not declared in this scope
##      return libewf_read_random(handle,buf,bytes,offset);

Some of these issues appear to be because the current libewf source no longer implements some of these functions as ___, but rather __handle__ (e.g. libewf_handle_open rather than libewf_open), and no longer provides the ability to support previous versions of the API.

**Comment by simsong** [comment]
How do you recommend we handle this?

**Comment by kamwoods** [comment]
My preference would be to update the bulk_extractor development source on the GitHub master to compile with libewf-experimental as it exists now, and drop all support for the legacy libewf calls.

Rationale follows. Please bear with me.

I spoke briefly via email with Joachim regarding a similar issue with TSK. The current TSK master doesn't build with libewf-experimental without a minor patch to ewf.c in the TSK sources (replacing libewf_handle_read_random with libewf_handle_read_buffer_at_offset). TSK 4.2.0 only officially supports a 2013 build of libewf if you want EWF support. But the "official" link on the TSK site goes to https://sourceforge.net/projects/libewf/, which no longer includes the libewf-20130416 build and points to the GitHub repo, which only has two recent libewf-experimental builds as releases, with which TSK fails to build (without the aforementioned minor patch)!

Joachim indicated his preference would be that everyone eventually move to the current sources and drop support for the v1 API, but that some additional changes might be needed in ewf.c in TSK to put everything just right.

Out in Linux distro-world, things are pretty much stable, since many distros have established packages for TSK 4.1.3 or 4.2.0, libewf-20130416, and bulk_extractor 1.5.3. 

But anyone who wants to compile TSK 4.2.0 (or current master sources), libewf-experimental, and bulk_extractor (1.6.0-dev) from scratch from the GitHub sources *right now* (using the provided instructions in the relevant READMEs) is going to run into the same problems I have. 

I don't believe that updating the master to work purely with libewf-experimental (breaking compatibility with earlier v1-API-capable libewf releases) will annoy anyone upstream, given that previous sources, releases, and working packages for bulk_extractor 1.5.3 are easily available, and that bulk_extractor will almost certainly need to updated eventually *anyway* to work with the current way libewf does things.

Of course, there may be concerns I've overlooked. This is just my current thinking.

**Comment by odibrowne** [comment]
Has there been any fixes to this issue. I am trying to compile bulk extractor with lightgrep but have the error codes similar to Kam
root@dft:~/Downloads/bulk_extractor# make
make  all-recursive
make[1]: Entering directory '/home/dft/Downloads/bulk_extractor'
Making all in doc
make[2]: Entering directory '/home/dft/Downloads/bulk_extractor/doc'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/home/dft/Downloads/bulk_extractor/doc'
Making all in doc/latex_manuals
make[2]: Entering directory '/home/dft/Downloads/bulk_extractor/doc/latex_manuals'
echo type \"make doc\" to create documentation
type "make doc" to create documentation
make[2]: Leaving directory '/home/dft/Downloads/bulk_extractor/doc/latex_manuals'
Making all in src
make[2]: Entering directory '/home/dft/Downloads/bulk_extractor/src'
depbase=`echo bulk_extractor_api.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H -I. -I..  -I../src/be13_api -I/usr/local/include -DUTC_OFFSET=+0000 -I/usr/local/include -I/usr/include/x86_64-linux-gnu  -g -pthread -g -O3 -std=c++11 -Wall -MD -D_FORTIFY_SOURCE=2 -Wpointer-arith -Wshadow -Wwrite-strings -Wcast-align -Wredundant-decls -Wdisabled-optimization -Wfloat-equal -Wmultichar -Wmissing-noreturn -Woverloaded-virtual -Wsign-promo -funit-at-a-time -Weffc++  -fPIC -D_FORTIFY_SOURCE=2  -MT bulk_extractor_api.o -MD -MP -MF $depbase.Tpo -c -o bulk_extractor_api.o bulk_extractor_api.cpp &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo base64_forensic.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H -I. -I..  -I../src/be13_api -I/usr/local/include -DUTC_OFFSET=+0000 -I/usr/local/include -I/usr/include/x86_64-linux-gnu  -g -pthread -g -O3 -std=c++11 -Wall -MD -D_FORTIFY_SOURCE=2 -Wpointer-arith -Wshadow -Wwrite-strings -Wcast-align -Wredundant-decls -Wdisabled-optimization -Wfloat-equal -Wmultichar -Wmissing-noreturn -Woverloaded-virtual -Wsign-promo -funit-at-a-time -Weffc++  -fPIC -D_FORTIFY_SOURCE=2  -MT base64_forensic.o -MD -MP -MF $depbase.Tpo -c -o base64_forensic.o base64_forensic.cpp &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo dig.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H -I. -I..  -I../src/be13_api -I/usr/local/include -DUTC_OFFSET=+0000 -I/usr/local/include -I/usr/include/x86_64-linux-gnu  -g -pthread -g -O3 -std=c++11 -Wall -MD -D_FORTIFY_SOURCE=2 -Wpointer-arith -Wshadow -Wwrite-strings -Wcast-align -Wredundant-decls -Wdisabled-optimization -Wfloat-equal -Wmultichar -Wmissing-noreturn -Woverloaded-virtual -Wsign-promo -funit-at-a-time -Weffc++  -fPIC -D_FORTIFY_SOURCE=2  -MT dig.o -MD -MP -MF $depbase.Tpo -c -o dig.o dig.cpp &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo findopts.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H -I. -I..  -I../src/be13_api -I/usr/local/include -DUTC_OFFSET=+0000 -I/usr/local/include -I/usr/include/x86_64-linux-gnu  -g -pthread -g -O3 -std=c++11 -Wall -MD -D_FORTIFY_SOURCE=2 -Wpointer-arith -Wshadow -Wwrite-strings -Wcast-align -Wredundant-decls -Wdisabled-optimization -Wfloat-equal -Wmultichar -Wmissing-noreturn -Woverloaded-virtual -Wsign-promo -funit-at-a-time -Weffc++  -fPIC -D_FORTIFY_SOURCE=2  -MT findopts.o -MD -MP -MF $depbase.Tpo -c -o findopts.o findopts.cpp &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo image_process.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H -I. -I..  -I../src/be13_api -I/usr/local/include -DUTC_OFFSET=+0000 -I/usr/local/include -I/usr/include/x86_64-linux-gnu  -g -pthread -g -O3 -std=c++11 -Wall -MD -D_FORTIFY_SOURCE=2 -Wpointer-arith -Wshadow -Wwrite-strings -Wcast-align -Wredundant-decls -Wdisabled-optimization -Wfloat-equal -Wmultichar -Wmissing-noreturn -Woverloaded-virtual -Wsign-promo -funit-at-a-time -Weffc++  -fPIC -D_FORTIFY_SOURCE=2  -MT image_process.o -MD -MP -MF $depbase.Tpo -c -o image_process.o image_process.cpp &&\
mv -f $depbase.Tpo $depbase.Po
image_process.cpp: In destructor ‘virtual process_ewf::~process_ewf()’:
image_process.cpp:333:21: error: ‘libewf_close’ was not declared in this scope
  libewf_close(handle);
                     ^
image_process.cpp: In member function ‘virtual int process_ewf::open()’:
image_process.cpp:456:98: error: too few arguments to function ‘int libewf_glob(const char*, size_t, uint8_t, char***, int*, libewf_error_t**)’
 mes = libewf_glob(fname,strlen(fname),LIBEWF_FORMAT_UNKNOWN,&libewf_filenames);
                                                                              ^
In file included from image_process.h:207:0,
                 from image_process.cpp:21:
/usr/local/include/libewf.h:134:5: note: declared here
 int libewf_glob(
     ^
image_process.cpp:460:79: error: ‘libewf_open’ was not declared in this scope
    handle = libewf_open(libewf_filenames,amount_of_filenames,LIBEWF_OPEN_READ);
                                                                              ^
image_process.cpp:468:59: error: ‘libewf_get_media_size’ was not declared in this scope
     libewf_get_media_size(handle,(size64_t *)&ewf_filesize);
                                                           ^
image_process.cpp: In member function ‘virtual int process_ewf::pread(unsigned char*, size_t, int64_t) const’:
image_process.cpp:520:54: error: ‘libewf_read_random’ was not declared in this scope
     return libewf_read_random(handle,buf,bytes,offset);
                                                      ^
Makefile:971: recipe for target 'image_process.o' failed
make[2]: *** [image_process.o] Error 1
make[2]: Leaving directory '/home/dft/Downloads/bulk_extractor/src'
Makefile:466: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/dft/Downloads/bulk_extractor'
Makefile:405: recipe for target 'all' failed
make: *** [all] Error 2
root@dft:~/Downloads/bulk_extractor# 

I am presently running ubuntu 16.04 and have installed the most recent version of libewf

**Comment by simsong** [comment]
Is this a lightgrep problem or a libewf problem?

**Comment by jonstewart** [comment]
This looks squarely like a `libewf` problem, as the error message is being generated from within `image_process.cpp` and only in relation to `libewf` calls.

Are you trying to run with experimental builds of `libewf`? I personally do not recommend this. My team found a big bug in some of the experimental releases (which we reported and is now fixed), and any arbitrary commit from its master branch has a nonzero probability of failing due to build problems with its many dependencies. I recommend you stick to the last stable release from legacy: https://github.com/libyal/legacy/.
